client/src/main/java/io/pravega/client/state/impl/ReaderGroupStateSynchronizer.java

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.state.impl;
+
+import io.pravega.client.SynchronizerClientFactory;
+import io.pravega.client.control.impl.Controller;
+import io.pravega.client.state.InitialUpdate;
+import io.pravega.client.state.StateSynchronizer;
+import io.pravega.client.state.SynchronizerConfig;
+import io.pravega.client.state.Update;
+import io.pravega.client.stream.Serializer;
+import io.pravega.client.stream.impl.ReaderGroupState;
+import io.pravega.client.stream.impl.SegmentWithRange;
+import lombok.val;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import static io.pravega.client.stream.impl.ReaderGroupImpl.getEndSegmentsForStreams;
+import static io.pravega.client.stream.impl.ReaderGroupImpl.getSegmentsForStreams;
+import static io.pravega.common.concurrent.Futures.getThrowingException;
+import static io.pravega.shared.NameUtils.getStreamForReaderGroup;
+
+/**
+ * A new {@link StateSynchronizer} implementation specifically with the {@link ReaderGroupState} to allow for special
+ * updates to the controller.
+ */
+public class ReaderGroupStateSynchronizer implements StateSynchronizer<ReaderGroupState> {
```

Comment:
Why do we need this new class?

Reply:
We cannot make an `updateStateUnconditionally` call with `ReaderGroupStateInit` within `StateSynchronizerImpl` as that class is parametrized to accept any generic state. So there is a new class `ReaderGroupStateSynchronizer` that specifically implements a `StateSynchronizer<ReaderGroupState>` so that we can make the readergroup state update in `fetchUpdates`.

Reply:
It does not look like this approach is "upgrade" proof.